src/nvim/eval/funcs.c

```diff
@@ -6314,6 +6314,97 @@ static void f_range(typval_T *argvars, typval_T *rettv, FunPtr fptr)
   }
 }
 
+/*
+ * Evaluate "expr" for readdir().
+ */
+static varnumber_T readdir_checkitem(typval_T *expr, const char *name)
+{
+  typval_T save_val;
+  typval_T rettv;
+  typval_T argv[2];
+  varnumber_T retval = 0;
+  bool error = false;
+
+  prepare_vimvar(VV_VAL, &save_val);
+  set_vim_var_string(VV_VAL, name, -1);
+  argv[0].v_type = VAR_STRING;
+  argv[0].vval.v_string = (char_u *)name;
+
+  if (eval_expr_typval(expr, argv, 1, &rettv) == FAIL)
+    goto theend;
+
+  retval = tv_get_number_chk(&rettv, &error);
+  if (error)
+    retval = -1;
+  tv_clear(&rettv);
+
+theend:
+  set_vim_var_string(VV_VAL, NULL, 0);
+  restore_vimvar(VV_VAL, &save_val);
+  return retval;
+}
+
+/*
+ * "readdir()" function
+ */
+static void f_readdir(typval_T *argvars, typval_T *rettv, FunPtr fptr)
+{
+  typval_T *expr;
+  const char *path;
+  garray_T ga;
+  Directory dir;
+
+  tv_list_alloc_ret(rettv, kListLenUnknown);
+  path = tv_get_string(&argvars[0]);
+  expr = &argvars[1];
+  ga_init(&ga, (int)sizeof(char *), 20);
+
+  if(!os_scandir(&dir, path)) {
+    smsg(_(e_notopen), path);
+  } else {
+    for (;;) {
+      int ignore;
+
+      path = os_scandir_next(&dir);
+      if (path == NULL)
+        break;
+
+      ignore = path[0] == '.' &&
+        (path[1] == NUL ||
+         (path[1] == '.' && path[2] == NUL));
```

Comment:
How should I write this lines? Now it looks as in Vim, but as for this is not very readable.